File: android/app/src/main/java/com/wootodo/sync/SyncModels.kt
package com.wootodo.sync

enum class DevicePlatform(val wireValue: String) {
    MACOS("macos"),
    ANDROID("android");

    companion object {
        fun fromWire(value: String): DevicePlatform = entries.firstOrNull { it.wireValue == value }
            ?: throw IllegalArgumentException("不支持的平台：$value")
    }
}

enum class PairingStatus(val wireValue: String) {
    OPEN("open"),
    CLAIMED("claimed"),
    CONFIRMED("confirmed"),
    EXPIRED("expired"),
    CANCELED("canceled");

    companion object {
        fun fromWire(value: String): PairingStatus = entries.firstOrNull { it.wireValue == value }
            ?: throw IllegalArgumentException("不支持的配对状态：$value")
    }
}

enum class SyncOperationKind(val wireValue: String) {
    UPSERT("upsert"),
    DELETE("delete"),
    COMPLETE("complete"),
    PASS("pass"),
    REORDER("reorder");

    companion object {
        fun fromWire(value: String): SyncOperationKind = entries.firstOrNull { it.wireValue == value }
            ?: throw IllegalArgumentException("不支持的同步操作：$value")
    }
}

data class EncryptedEnvelope(
    val ciphertext: String,
    val nonce: String,
)

data class DeviceRegistration(
    val name: String,
    val platform: DevicePlatform,
    val publicKey: String? = null,
)

data class CreateVaultRequest(
    val device: DeviceRegistration,
    val recoveryEnvelope: EncryptedEnvelope? = null,
)

data class CreatedDevice(
    val id: String,
    val name: String,
    val platform: DevicePlatform,
    val token: String,
)

data class CreateVaultData(
    val vaultId: String,
    val device: CreatedDevice,
    val serverTime: Long,
)

data class CreatePairingRequest(val publicKey: String)

data class CreatePairingData(
    val pairingId: String,
    val pairingSecret: String,
    val initiatorPublicKey: String,
    val expiresAt: Long,
    val serverTime: Long,
)

data class PairingClaimRequest(
    val pairingSecret: String,
    val deviceToken: String,
    val device: DeviceRegistration,
)

data class PairingClaimData(
    val pairingId: String,
    val status: PairingStatus,
    val deviceId: String,
    val expiresAt: Long,
)

data class PairingClaimInfo(
    val deviceId: String,
    val name: String,
    val platform: DevicePlatform,
    val publicKey: String,
    val claimedAt: Long,
)

data class PairingStatusData(
    val pairingId: String,
    val status: PairingStatus,
    val expiresAt: Long,
    val claim: PairingClaimInfo?,
)

data class PairingConfirmRequest(val vaultKeyEnvelope: EncryptedEnvelope)

data class PairingConfirmData(
    val pairingId: String,
    val status: PairingStatus,
    val deviceId: String,
)

data class PairingResultRequest(
    val pairingSecret: String,
    val deviceToken: String,
)

data class PairingResultData(
    val pairingId: String,
    val status: PairingStatus,
    val vaultId: String?,
    val deviceId: String?,
    val initiatorPublicKey: String?,
    val vaultKeyEnvelope: EncryptedEnvelope?,
    val expiresAt: Long,
)

data class SyncPushOperation(
    val opId: String,
    val entityId: String,
    val kind: SyncOperationKind,
    val lamport: Long,
    val ciphertext: String,
    val nonce: String,
)

data class SyncRequest(
    val cursor: Long,
    val ack: Long? = null,
    val pullLimit: Int? = null,
    val push: List<SyncPushOperation>,
)

data class SyncPushSummary(
    val received: Int,
    val inserted: Int,
    val duplicates: Int,
)

data class SyncPulledOperation(
    val serverSeq: Long,
    val opId: String,
    val deviceId: String,
    val entityId: String,
    val kind: SyncOperationKind,
    val lamport: Long,
    val ciphertext: String,
    val nonce: String,
    val createdAt: Long,
) {
    fun metadata(): SyncOperationMetadata = SyncOperationMetadata(
        opId = opId,
        entityId = entityId,
        kind = kind,
        lamport = lamport,
        deviceId = deviceId,
    )
}

data class SyncData(
    val push: SyncPushSummary,
    val pull: List<SyncPulledOperation>,
    val cursor: Long,
    val hasMore: Boolean,
    val serverTime: Long,
)

data class DeviceInfo(
    val id: String,
    val name: String,
    val platform: DevicePlatform,
    val publicKey: String?,
    val createdAt: Long,
    val lastSeenAt: Long?,
    val revokedAt: Long?,
    val isCurrent: Boolean,
)

data class DeviceListData(val devices: List<DeviceInfo>)

data class RevokeDeviceData(
    val deviceId: String,
    val revokedAt: Long,
)

sealed interface JsonValue {
    data class Text(val value: String) : JsonValue
    data class NumberValue(val value: Double) : JsonValue
    data class BooleanValue(val value: Boolean) : JsonValue
    data class ObjectValue(val value: Map<String, JsonValue>) : JsonValue
    data class ArrayValue(val value: List<JsonValue>) : JsonValue
    data object NullValue : JsonValue
}

data class ServerErrorPayload(
    val code: String,
    val message: String,
    val details: Map<String, JsonValue>? = null,
)

data class BearerCredential(val deviceToken: String) {
    init {
        require(Base64Url.decode(deviceToken).size == 32) { "设备令牌必须是 32 字节" }
    }
}

object SyncProtocolLimits {
    const val MAX_PUSH_OPERATIONS = 50
    const val MAX_PULL_OPERATIONS = 100
    const val MAX_CIPHERTEXT_BYTES = 32 * 1024
}

File: macos/Sources/WooTodoSync/SyncModels.swift
import Foundation

public enum DevicePlatform: String, Codable, Sendable {
    case macos
    case android
}

public enum PairingStatus: String, Codable, Sendable {
    case open
    case claimed
    case confirmed
    case expired
    case canceled
}

public enum SyncOperationKind: String, Codable, CaseIterable, Sendable {
    case upsert
    case delete
    case complete
    case pass
    case reorder
}

public struct EncryptedEnvelope: Codable, Equatable, Sendable {
    public let ciphertext: String
    public let nonce: String

    public init(ciphertext: String, nonce: String) {
        self.ciphertext = ciphertext
        self.nonce = nonce
    }
}

public struct DeviceRegistration: Codable, Equatable, Sendable {
    public let name: String
    public let platform: DevicePlatform
    public let publicKey: String?

    public init(name: String, platform: DevicePlatform, publicKey: String? = nil) {
        self.name = name
        self.platform = platform
        self.publicKey = publicKey
    }
}

public struct CreateVaultRequest: Codable, Equatable, Sendable {
    public let device: DeviceRegistration
    public let recoveryEnvelope: EncryptedEnvelope?

    public init(device: DeviceRegistration, recoveryEnvelope: EncryptedEnvelope? = nil) {
        self.device = device
        self.recoveryEnvelope = recoveryEnvelope
    }
}

public struct CreatedDevice: Codable, Equatable, Sendable {
    public let id: String
    public let name: String
    public let platform: DevicePlatform
    public let token: String
}

public struct CreateVaultData: Codable, Equatable, Sendable {
    public let vaultId: String
    public let device: CreatedDevice
    public let serverTime: Int64
}

public struct CreatePairingRequest: Codable, Equatable, Sendable {
    public let publicKey: String

    public init(publicKey: String) {
        self.publicKey = publicKey
    }
}

public struct CreatePairingData: Codable, Equatable, Sendable {
    public let pairingId: String
    public let pairingSecret: String
    public let initiatorPublicKey: String
    public let expiresAt: Int64
    public let serverTime: Int64
}

public struct PairingClaimRequest: Codable, Equatable, Sendable {
    public let pairingSecret: String
    public let deviceToken: String
    public let device: PairingDeviceRegistration

    public init(pairingSecret: String, deviceToken: String, device: PairingDeviceRegistration) {
        self.pairingSecret = pairingSecret
        self.deviceToken = deviceToken
        self.device = device
    }
}

public struct PairingDeviceRegistration: Codable, Equatable, Sendable {
    public let name: String
    public let platform: DevicePlatform
    public let publicKey: String

    public init(name: String, platform: DevicePlatform, publicKey: String) {
        self.name = name
        self.platform = platform
        self.publicKey = publicKey
    }
}

public struct PairingClaimData: Codable, Equatable, Sendable {
    public let pairingId: String
    public let status: PairingStatus
    public let deviceId: String
    public let expiresAt: Int64
}

public struct PairingClaimInfo: Codable, Equatable, Sendable {
    public let deviceId: String
    public let name: String
    public let platform: DevicePlatform
    public let publicKey: String
    public let claimedAt: Int64
}

public struct PairingStatusData: Codable, Equatable, Sendable {
    public let pairingId: String
    public let status: PairingStatus
    public let expiresAt: Int64
    public let claim: PairingClaimInfo?
}

public struct PairingConfirmRequest: Codable, Equatable, Sendable {
    public let vaultKeyEnvelope: EncryptedEnvelope

    public init(vaultKeyEnvelope: EncryptedEnvelope) {
        self.vaultKeyEnvelope = vaultKeyEnvelope
    }
}

public struct PairingConfirmData: Codable, Equatable, Sendable {
    public let pairingId: String
    public let status: PairingStatus
    public let deviceId: String
}

public struct PairingResultRequest: Codable, Equatable, Sendable {
    public let pairingSecret: String
    public let deviceToken: String

    public init(pairingSecret: String, deviceToken: String) {
        self.pairingSecret = pairingSecret
        self.deviceToken = deviceToken
    }
}

public struct PairingResultData: Codable, Equatable, Sendable {
    public let pairingId: String
    public let status: PairingStatus
    public let vaultId: String?
    public let deviceId: String?
    public let initiatorPublicKey: String?
    public let vaultKeyEnvelope: EncryptedEnvelope?
    public let expiresAt: Int64

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.pairingId = try container.decode(String.self, forKey: .pairingId)
        self.status = try container.decode(PairingStatus.self, forKey: .status)
        self.vaultId = try container.decodeIfPresent(String.self, forKey: .vaultId)
        self.deviceId = try container.decodeIfPresent(String.self, forKey: .deviceId)
        self.initiatorPublicKey = try container.decodeIfPresent(String.self, forKey: .initiatorPublicKey)
        self.vaultKeyEnvelope = try container.decodeIfPresent(
            EncryptedEnvelope.self,
            forKey: .vaultKeyEnvelope
        )
        self.expiresAt = try container.decode(Int64.self, forKey: .expiresAt)

        switch status {
        case .claimed:
            guard vaultId == nil, deviceId == nil, initiatorPublicKey == nil,
                  vaultKeyEnvelope == nil else {
                throw DecodingError.dataCorruptedError(
                    forKey: .status,
                    in: container,
                    debugDescription: "claimed 配对结果不得携带确认字段"
                )
            }
        case .confirmed:
            guard vaultId != nil, deviceId != nil, initiatorPublicKey != nil,
                  vaultKeyEnvelope != nil else {
                throw DecodingError.dataCorruptedError(
                    forKey: .status,
                    in: container,
                    debugDescription: "confirmed 配对结果缺少确认字段"
                )
            }
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .status,
                in: container,
                debugDescription: "配对 result 只允许 claimed 或 confirmed"
            )
        }
    }
}

public struct SyncPushOperation: Codable, Equatable, Sendable {
    public let opId: String
    public let entityId: String
    public let kind: SyncOperationKind
    public let lamport: Int64
    public let ciphertext: String
    public let nonce: String

    public init(
        opId: String,
        entityId: String,
        kind: SyncOperationKind,
        lamport: Int64,
        ciphertext: String,
        nonce: String
    ) {
        self.opId = opId
        self.entityId = entityId
        self.kind = kind
        self.lamport = lamport
        self.ciphertext = ciphertext
        self.nonce = nonce
    }
}

public struct SyncRequest: Codable, Equatable, Sendable {
    public let cursor: Int64
    public let ack: Int64?
    public let pullLimit: Int?
    public let push: [SyncPushOperation]

    public init(
        cursor: Int64,
        ack: Int64? = nil,
        pullLimit: Int? = nil,
        push: [SyncPushOperation]
    ) {
        self.cursor = cursor
        self.ack = ack
        self.pullLimit = pullLimit
        self.push = push
    }
}

public struct SyncPushSummary: Codable, Equatable, Sendable {
    public let received: Int
    public let inserted: Int
    public let duplicates: Int

    public init(received: Int, inserted: Int, duplicates: Int) {
        self.received = received
        self.inserted = inserted
        self.duplicates = duplicates
    }
}

public struct SyncPulledOperation: Codable, Equatable, Sendable {
    public let serverSeq: Int64
    public let opId: String
    public let deviceId: String
    public let entityId: String
    public let kind: SyncOperationKind
    public let lamport: Int64
    public let ciphertext: String
    public let nonce: String
    public let createdAt: Int64

    public init(
        serverSeq: Int64,
        opId: String,
        deviceId: String,
        entityId: String,
        kind: SyncOperationKind,
        lamport: Int64,
        ciphertext: String,
        nonce: String,
        createdAt: Int64
    ) {
        self.serverSeq = serverSeq
        self.opId = opId
        self.deviceId = deviceId
        self.entityId = entityId
        self.kind = kind
        self.lamport = lamport
        self.ciphertext = ciphertext
        self.nonce = nonce
        self.createdAt = createdAt
    }
}

public struct SyncData: Codable, Equatable, Sendable {
    public let push: SyncPushSummary
    public let pull: [SyncPulledOperation]
    public let cursor: Int64
    public let hasMore: Bool
    public let serverTime: Int64

    public init(
        push: SyncPushSummary,
        pull: [SyncPulledOperation],
        cursor: Int64,
        hasMore: Bool,
        serverTime: Int64
    ) {
        self.push = push
        self.pull = pull
        self.cursor = cursor
        self.hasMore = hasMore
        self.serverTime = serverTime
    }
}

public struct DeviceInfo: Codable, Equatable, Sendable {
    public let id: String
    public let name: String
    public let platform: DevicePlatform
    public let publicKey: String?
    public let createdAt: Int64
    public let lastSeenAt: Int64?
    public let revokedAt: Int64?
    public let isCurrent: Bool
}

public struct DeviceListData: Codable, Equatable, Sendable {
    public let devices: [DeviceInfo]
}

public struct RevokeDeviceData: Codable, Equatable, Sendable {
    public let deviceId: String
    public let revokedAt: Int64
}

public enum JSONValue: Codable, Equatable, Sendable {
    case string(String)
    case number(Double)
    case boolean(Bool)
    case object([String: JSONValue])
    case array([JSONValue])
    case null

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let value = try? container.decode(Bool.self) {
            self = .boolean(value)
        } else if let value = try? container.decode(Double.self) {
            self = .number(value)
        } else if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode([String: JSONValue].self) {
            self = .object(value)
        } else if let value = try? container.decode([JSONValue].self) {
            self = .array(value)
        } else {
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "无法解析服务端错误详情"
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let value): try container.encode(value)
        case .number(let value): try container.encode(value)
        case .boolean(let value): try container.encode(value)
        case .object(let value): try container.encode(value)
        case .array(let value): try container.encode(value)
        case .null: try container.encodeNil()
        }
    }
}

public struct ServerErrorPayload: Codable, Equatable, Sendable {
    public let code: String
    public let message: String
    public let details: [String: JSONValue]?

    public init(code: String, message: String, details: [String: JSONValue]? = nil) {
        self.code = code
        self.message = message
        self.details = details
    }
}
